### ISSUE TITLE
Pipeline not removed when a space is removed

### ISSUE THREAD

**Issue by ppitonak** [openshiftio/openshift.io#3802]
1. create a new project
2. wait for pipeline to create
3. verify on https://jenkins.openshift.io/job/yourusername that the pipeline really exists
4. go to My Spaces
5. remove the space
6. verify on https://jenkins.openshift.io/job/yourusername that the pipeline was removed

expected:
* pipeline is not listed in Jenkins UI

result:
* pipeline was not removed
* pipeline will be removed no matter what state it is in (new/pending/running/failed)

**Comment by joshuawilson** [comment]
@aslakknutsen @xcoulon is this covered by the work to do a complete delete of everything? is it a duplicate?

**Comment by stevengutz** [comment]
I don't think this is a SEV2 - nothing really failed here.  Just a process lingering that should have been killed.

**Comment by joshuawilson** [comment]
You are unable to remove a pipeline. The pipeline will continue to take up resources. That is why I marked it a SEV2. If you or the build team (or anyone else) wants to change the label, they may do that.

**Comment by ppitonak** [comment]
As @joshuawilson wrote, it's using resources so I agree with SEV2 label.

**Comment by piyush1594** [comment]
Is build team need to do anything related to this. @ppitonak @joshuawilson cc @sthaha

**Comment by sthaha** [comment]
@piyush1594 yes, this is a build-cd bug as we should be cleaning up unwanted resources.
IMHO the service that is creating the job/pipeline must be the one cleanup it up as well.  So the question is what is creating that job? 

I also wonder if this issue is an indirect cause of cpu/memory spike when jenkins start.

**Comment by rupalibehera** [comment]
This is a build-cd bug, we should consider this issue in this or next sprint

**Comment by sthaha** [comment]
@rupalibehera yes, I have added the `sprint/next` label and is in our backlog.

**Comment by lordofthejars** [comment]
@rupalibehera can you write down your conclusions so I can fix it? Or you prefer to fix by yourself since you already started to investigate?

**Comment by rupalibehera** [comment]
@lordofthejars, I have not investigated on this issue.

**Comment by rupalibehera** [comment]
@lordofthejars , 
1. Check if we can reproduce this issue
2. Launcher creates the configmap here https://github.com/fabric8-launcher/launcher-backend/blob/763937f8186359d6a7732f066592e880621d5b74/addons/osio-addon/src/main/java/io/fabric8/launcher/osio/OsioMissionControl.java#L156, then the jenkins sync plugin uses the configmap to create jobs 
3. I am not sure how the deletion is handled, we need to check in that case, may be check if something you can find in sync plugin or launcher, if launcher then we can assign the issue to launcher.

**Comment by lordofthejars** [comment]
@gastaldi @rupalibehera @sthaha @ppitonak I have been investigating this. And I have noticed that it is reproducible.

But I have noticed another thing and it is that if I login into Jenkins and go to Jenkins admin console and I tried to delete the job manually, but I have noticed that there is no `Delete Pipeline` button.

So it seems that from Jenkins point of view the user has no rights for deleting a job, so this could be a problem.

@gastaldi I think that you are also into the launcher, any thoughts about my thoughts?

**Comment by gastaldi** [comment]
@lordofthejars deleting pipelines or jobs (or any other resource) is out of scope of Launcher.

**Comment by lordofthejars** [comment]
@gastaldi ok then when a space is deleted, where is the logic which receives the call of delete space and deals with it? My thoughts are that since it is a delete they are calling delete Job to Jenkins(or whatever) it fails because of permission but the error is ignored.

**Comment by gastaldi** [comment]
@lordofthejars I have no idea, maybe someone from the platform team may be able to answer that (@aslakknutsen @joshuawilson  ?)

**Comment by lordofthejars** [comment]
This is not an easy task, the first thing we need to do is create a build API which offers, apart from other operations,  a `delete` operation to external callers.

Then when space is deleted, this method should be called.

**Comment by sthaha** [comment]
@gastaldi 
> @lordofthejars deleting pipelines or jobs (or any other resource) is out of scope of Launcher.

I agree with it in principle but since launcher creates these pipelines, wouldn't it make sense for the service that creates it to also cleanup? Or would rather suggest that creation of jobs and pipeline must move out of launcher to some service that should be under build? 

@lordofthejars 
> This is not an easy task, the first thing we need to do is create a build API which offers, apart from other operations, a delete operation to external callers.

Thanks for the investigation. I agree it is a non-trivial amount of work.

**Comment by gastaldi** [comment]
>creation of jobs and pipeline must move out of launcher to some service that should be under build?

Yes, that would be the ideal choice. IMHO in a perfect architecture, Launcher would send an event to a bus saying that a GitHub repository was created and the interested parties would react accordingly.

IMHO Launcher shouldn't be doing more than just what was designed to do (launching a booster).

**Comment by hrishin** [comment]
Above discussion makes sense. On other hand do we ETA for event-based implementation?
If its gonna take long meanwhile it could be fixed in the platform. Shall we go for it?

**Comment by joshuawilson** [comment]
I've also added this to a story to track the e2e bugs at https://openshift.io/openshiftio/Openshift_io/plan/detail/802

**Comment by alexeykazakov** [comment]
@ppitonak why it's a blocker for e2e tests? I agree that it's not optimal right now. We should do a better clean up when deleting a space. There are other similar issues when we don't remove WIs for example when deleting a space but it doesn't break anything, does it?

**Comment by ppitonak** [comment]
@alexeykazakov we will remove the `area/e2e-tests` label for now.

For the record, this issue doesn't block us at the moment therefore we don't consider it to be a blocker. However, it's important to fix it in the future because we would like to switch from resetting the account to removing space in our test.

**Comment by hrishin** [comment]
### Update:

For the time being, we are adding this functionality into `fabric8-wit`, WIP https://github.com/hrishin/fabric8-wit/tree/delete-pipeline.

Eventually we have to move this to  `build-service`. 
cc: @chmouel

**Comment by hrishin** [comment]
### Update:

Started working on this the PR, will update it by tomorrow. 
Thanks!
